zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs

```diff
@@ -0,0 +1,371 @@
+use eth_types::{Field, ToLittleEndian, ToScalar};
+use gadgets::util::Expr;
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::N_BYTES_ACCOUNT_ADDRESS,
+        step::ExecutionState,
+        util::{
+            common_gadget::RestoreContextGadget,
+            constraint_builder::{ConstrainBuilderCommon, EVMConstraintBuilder},
+            from_bytes, rlc, CachedRegion, Cell, RandomLinearCombination,
+        },
+    },
+    table::CallContextFieldTag,
+    witness::{Block, Call, ExecStep, Transaction},
+};
+
+#[derive(Clone, Debug)]
+pub struct EcrecoverGadget<F> {
+    recovered: Cell<F>,
+    msg_hash_rlc: Cell<F>,
+    sig_v_rlc: Cell<F>,
+    sig_r_rlc: Cell<F>,
+    sig_s_rlc: Cell<F>,
+    recovered_addr_rlc: RandomLinearCombination<F, N_BYTES_ACCOUNT_ADDRESS>,
+
+    is_success: Cell<F>,
+    callee_address: Cell<F>,
+    caller_id: Cell<F>,
+    call_data_offset: Cell<F>,
+    call_data_length: Cell<F>,
+    return_data_offset: Cell<F>,
+    return_data_length: Cell<F>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for EcrecoverGadget<F> {
+    const EXECUTION_STATE: ExecutionState = ExecutionState::PrecompileEcrecover;
+
+    const NAME: &'static str = "ECRECOVER";
+
+    fn configure(cb: &mut EVMConstraintBuilder<F>) -> Self {
+        let (recovered, msg_hash_rlc, sig_v_rlc, sig_r_rlc, sig_s_rlc, recovered_addr_rlc) = (
+            cb.query_bool(),
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+            cb.query_cell_phase2(),
+            cb.query_keccak_rlc(),
+        );
+
+        cb.condition(recovered.expr(), |cb| {
+            // if address was recovered, the sig_v (recovery ID) was correct.
+            cb.require_zero(
+                "sig_v == 27 or 28",
+                (sig_v_rlc.expr() - 27.expr()) * (sig_v_rlc.expr() - 28.expr()),
+            );
+
+            // lookup to the sign_verify table
+            // || v | r | s | msg_hash | recovered_addr ||
+            cb.sig_table_lookup(
+                msg_hash_rlc.expr(),
+                sig_v_rlc.expr() - 27.expr(),
+                sig_r_rlc.expr(),
+                sig_s_rlc.expr(),
+                from_bytes::expr(&recovered_addr_rlc.cells),
+            );
+        });
+
+        let [is_success, callee_address, caller_id, call_data_offset, call_data_length, return_data_offset, return_data_length] =
+            [
+                CallContextFieldTag::IsSuccess,
+                CallContextFieldTag::CalleeAddress,
+                CallContextFieldTag::CallerId,
+                CallContextFieldTag::CallDataOffset,
+                CallContextFieldTag::CallDataLength,
+                CallContextFieldTag::ReturnDataOffset,
+                CallContextFieldTag::ReturnDataLength,
+            ]
+            .map(|tag| cb.call_context(None, tag));
```

Comment:
`caller_id, call_data_offset, call_data_length, return_data_offset, return_data_length` are not used elsewhere. So here we only read the call_context.   Is it needed really?